File: #0053.maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {

        vector<int> v(nums.size());
        v[0] = nums[0];
        int size = nums.size();
        for (int i = 1; i < size; ++i) {
            v[i] = max(v[i - 1] + nums[i], nums[i]);
        }

        return *std::max_element(v.begin(), v.end());
    }
};

File: #0053.maximum-subarray.rs
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {

        nums.iter().fold((0, std::i32::MIN), |(it, max), &num| {
            (std::cmp::max(0, it + num) ,std::cmp::max(max, it + num))
        }).1
    }
}
